gensim/corpora/textcorpus.py

```diff
@@ -409,66 +516,130 @@ def getstream(self):
-    for name in names:
-        if isdir(join(top, name)):
-            dirs.append(name)
-        else:
-            nondirs.append(name)
 
-    if topdown:
-        yield depth, top, dirs, nondirs
+    def __init__(self, source, max_sentence_length=MAX_WORDS_IN_BATCH, limit=None, **kwargs):
+        """
+        `source` can be either a string or a file object. Clip the file to the first
+        `limit` lines (or no clipped if limit is None, the default).
+
+        Example::
+
+            sentences = LineSentence('myfile.txt')
+
+        Or for compressed files::
+
+            sentences = LineSentence('compressed_text.txt.bz2')
+            sentences = LineSentence('compressed_text.txt.gz')
+
+        """
+        self.max_sentence_length = max_sentence_length
+        self.limit = limit
+        kwargs['tokenizer'] = kwargs.get('tokenizer', unicode_and_tokenize)
+        kwargs['character_filters'] = kwargs.get('character_filters', [])
+        kwargs['token_filters'] = kwargs.get('token_filters', [])
+        kwargs['processes'] = kwargs.get('processes', 1)
+        TextCorpus.__init__(self, source, **kwargs)
+
+    def getstream(self):
+        with utils.smart_open(self.source) as fin:
+            for line in itertools.islice(fin, self.limit):
+                yield line
 
-    # Again O(n), where n = number of directories in the directory
-    for name in dirs:
-        new_path = join(top, name)
-        if followlinks or not islink(new_path):
+    def yield_tokens(self):
+        doc_token_stream = super(self.__class__, self).yield_tokens()
+        for tokens in doc_token_stream:
+            i = 0
+            while i < len(tokens):
+                yield tokens[i: i + self.max_sentence_length]
+                i += self.max_sentence_length
 
-            # Generator so besides the recursive `walk()` call, no additional cost here.
-            for x in walk(new_path, topdown, onerror, followlinks, depth + 1):
-                yield x
-    if not topdown:
-        yield depth, top, dirs, nondirs
+
+class Text8Corpus(TextTokensIterator, TextCorpus):
+    """Iterate over sentences from the "text8" corpus,
+    unzipped from http://mattmahoney.net/dc/text8.zip.
+    """
+
+    def __init__(self, source, max_sentence_length=MAX_WORDS_IN_BATCH, chunksize=65536, **kwargs):
+        self.max_sentence_length = max_sentence_length
+        self.chunksize = chunksize
+        kwargs['tokenizer'] = kwargs.get('tokenizer', unicode_and_tokenize)
+        kwargs['character_filters'] = kwargs.get('character_filters', [])
+        TextCorpus.__init__(self, source, **kwargs)
+
+    def _sentence_token_stream(self):
+        # Entire corpus is one gigantic line -- there are no sentence marks at all.
+        # So just split the token sequence arbitrarily into sentences of length
+        # `max_sentence_length`.
+        sentence, rest = [], b''
+        with utils.smart_open(self.source) as fin:
```

Comment:
Best to open files in binary mode (`rb`), and convert to text explicitly where needed.